crates/cadenza-eval/STATUS.md

```diff
@@ -684,6 +684,98 @@ The following enhancements were identified during code review and are planned fo
 - [ ] **Native targets**: Emit Rust code (primary), Cranelift/LLVM (optional)
 - [ ] **Dead code elimination**: Remove unused specializations
 
+#### WASM Code Generation
+
+**Status**: Initial implementation merged with basic WAT output generation. Critical next step is proper value management.
+
+**Current State**:
+- [x] Basic WASM codegen infrastructure with wasm-encoder
+- [x] WAT snapshot testing for all test-data files
+- [x] Function type signatures and declarations
+- [x] Simple constant generation (integers, floats, booleans)
+- [x] Binary operations (arithmetic, comparison) - instruction generation only
+- [x] Module structure with proper section ordering
+- [x] ~~Value location tracking with local management~~ ✅ (Completed - all 89 WAT tests pass)
+
+**Critical Issues** (blocking correct WASM output):
+1. ~~**Value Location Tracking**~~: ✅ **FIXED** - WASM codegen now properly tracks SSA values as WASM locals and generates correct `local.get`/`local.set` instructions.
+
+**High Priority Tasks** (in order):
+- [x] ~~**Value location tracking and local management**~~  ✅
+  - [x] Design data structure to map ValueId → LocalIndex or StackPosition
+  - [x] Track function parameters as locals 0..N
+  - [x] Allocate locals for SSA values (one local per ValueId that needs storage)
+  - [x] Implement strategy: parameters as locals, results pushed to stack
+  - [x] Generate `local.get` instructions to load operands before operations
+  - [x] Generate `local.set` instructions when values need to be stored
+  - [x] Handle the case where a value is used multiple times (must store in local)
+- [ ] **Function calls**
+  - [ ] Load arguments in order using `local.get` or stack values
+  - [ ] Generate `call` instruction with correct function index
+  - [ ] Handle return values on stack
+  - [ ] Test with recursive functions
+- [ ] **Control flow**
+  - [ ] Implement branch instruction generation (conditional jumps)
+  - [ ] Implement unconditional jump (br) for loops
+  - [ ] Handle block nesting for structured control flow
+  - [ ] Generate proper WASM blocks and loops
+  - [ ] Map IR basic blocks to WASM control structures
+- [ ] **Unary operations**
+  - [ ] Fix negation to properly load operand first
+  - [ ] Fix logical not with proper type conversions
+  - [ ] Implement bitwise not
+
+**Medium Priority** (after basic codegen works):
+- [ ] **String constants**
+  - [ ] Use WASM GC arrays for string representation
+  - [ ] Import/export string handling functions
+  - [ ] Consider WTF-8 encoding for compatibility
+- [ ] **Quantity/measurement values**
+  - [ ] Represent as structs with value and unit fields
+  - [ ] Or compile-time eliminate units and use raw numbers
+- [ ] **Records and tuples**
```

Comment:
Yes they should use the same structure. Tuples are just records with number fields instead of idents. But I think by the time we get to codegen the IR should eliminate tuples and we'd just be left with records so there's one less thing to worry about.